File: rust/src/leetcode/Q746_min_cost_climbing_stairs.rs
/*
 * @lc app=leetcode id=746 lang=rust
 *
 * [746] Min Cost Climbing Stairs
 */

// @lc code=start
impl Solution {
    pub fn min_cost_climbing_stairs(cost: Vec<i32>) -> i32 {
        let mut steps: Vec<i32> = vec![0; cost.len() + 1];

        for i in 2..=cost.len() {
            steps[i] = (steps[i - 1] + cost[i - 1]).min(steps[i - 2] + cost[i - 2]);
        }

        steps[steps.len() - 1]
    }
}
// @lc code=end

File: python/Q746_min_cost_climbing_stairs.py
#
# @lc app=leetcode id=746 lang=python3
#
# [746] Min Cost Climbing Stairs
#


# @lc code=start
class Solution:
    def minCostClimbingStairs(self, cost: list[int]) -> int:
        n: int = len(cost)
        steps: list[int] = [0] * n
        steps[n - 1] = cost[n - 1]
        steps[n - 2] = cost[n - 2]

        for i in range(n - 3, -1, -1):
            steps[i] = min(steps[i + 1], steps[i + 2]) + cost[i]

        return min(steps[0], steps[1])


# @lc code=end
